File: scala/src/main/scala/graph_degeneracy.scala
import scala.collection.mutable

class degeneracy_ordering[Vertex: Integral] private (
    graph: UndirectedGraph[Vertex],
    priority_per_vertex: Array[Int],
    queue: PriorityQueue[Vertex],
    var num_left_to_pick: Int
) extends Iterator[Vertex] {
  // priority_per_vertex:
  // If priority is 0, vertex was already picked or was always irrelevant (unconnected);
  // otherwise, vertex is still queued and priority = degree + 1 - number of picked neighbours.

  private def priority_of_vertex(v: Vertex): Int = {
    priority_per_vertex(implicitly[Integral[Vertex]].toInt(v))
  }

  private def prioritize_vertex(v: Vertex, p: Int): Unit = {
    priority_per_vertex(implicitly[Integral[Vertex]].toInt(v)) = p
  }

  override def hasNext: Boolean = { num_left_to_pick > 0 }

  override def next(): Vertex = {
    assert(
      priority_per_vertex.zipWithIndex
        .map { case (p, v) => (p, implicitly[Integral[Vertex]].fromInt(v)) }
        .forall { case (p, v) => p == 0 || queue.contains(p, v) }
    )
    var i = queue.pop().get
    while (priority_of_vertex(i) == 0) {
      // v was requeued with a more urgent priority and therefore already picked
      i = queue.pop().get
    }

    prioritize_vertex(i, 0)
    for (v <- graph.neighbours(i)) {
      val old_priority = priority_of_vertex(v)
      if (old_priority != 0) {
        // Since this is an unvisited neighbour of a vertex just being picked,
        // its priority can't be down to the minimum.
        val new_priority = old_priority - 1
        assert(new_priority > 0)
        // Requeue with a more urgent priority, but don't bother to remove
        // the original entry - it will be skipped if it's reached at all.
        prioritize_vertex(v, new_priority)
        queue.put(new_priority, v)
      }
    }
    num_left_to_pick -= 1
    i
  }
}

object degeneracy_ordering {
  def apply[Vertex: Integral](
      graph: UndirectedGraph[Vertex],
      drop: Int
  ): degeneracy_ordering[Vertex] = {
    assert(drop <= 0)
    val order = graph.order()
    var max_priority = 0
    val priority_per_vertex = Array.fill[Int](order)(0)
    var num_candidates = 0
    for (i <- 0 until order) {
      val c = implicitly[Integral[Vertex]].fromInt(i)
      val degree = graph.degree(c)
      if (degree > 0) {
        val priority = degree + 1
        max_priority = math.max(max_priority, priority)
        priority_per_vertex(i) = priority
        num_candidates += 1
      }
    }
    val queue = new PriorityQueue[Vertex](max_priority)
    for (i <- 0 until order) {
      val c = implicitly[Integral[Vertex]].fromInt(i)
      val priority = priority_per_vertex(i)
      if (priority != 0) {
        queue.put(priority, c)
      }
    }
    new degeneracy_ordering(
      graph,
      priority_per_vertex,
      queue,
      num_candidates + drop
    )
  }
}

class PriorityQueue[Vertex](max_priority: Int) {
  val stack_per_priority: Array[mutable.ArrayBuffer[Vertex]] =
    Array.fill(max_priority) { new mutable.ArrayBuffer[Vertex] }

  def put(priority: Int, v: Vertex): Unit = {
    stack_per_priority(priority - 1) += v
  }

  def pop(): Option[Vertex] = {
    for (stack <- stack_per_priority) {
      val element = stack.lastOption
      if (element.isDefined) {
        stack.remove(stack.length - 1)
        return element
      }
    }
    None
  }

  def contains(priority: Int, v: Vertex): Boolean = {
    stack_per_priority(priority - 1).contains(v)
  }
}

File: python3/graph_degeneracy.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from typing import Generator, List


class Degeneracy:
    def __init__(self, graph: UndirectedGraph) -> None:
        self.graph = graph
        # Possible values of priorityPerVertex (after initialization):
        #   0: never queued because not connected (degree 0),
        #      or no longer queued because it has been yielded itself,
        #      or no longer queued because all neighbours have been yielded
        #   1 or more: candidates queued with priority (degree - #of yielded neighbours)
        self.priority_per_node = [0] * graph.order
        self.q = PriorityQueue(max_priority=graph.max_degree)
        self.num_left_to_pick = 0
        for v in range(graph.order):
            if priority := graph.degree(v):
                self.priority_per_node[v] = priority
                self.q.put(priority=priority, element=v)
                self.num_left_to_pick += 1

    def is_candidate(self, v: Vertex) -> bool:
        return self.priority_per_node[v] != 0

    def iter(self) -> Generator[Vertex, None, None]:
        """
        Iterate connected vertices, lowest degree first, skipping vertices whose neighbours
        have all been yielded.
        """
        while self.num_left_to_pick > 0:
            pick = self.q.pop()
            if self.priority_per_node[pick] > 0:
                self.priority_per_node[pick] = 0
                neighbours = self.graph.adjacencies[pick]
                self.num_left_to_pick -= 1
                yield pick
                for v in neighbours:
                    if (old_priority := self.priority_per_node[v]) > 0:
                        # Requeue with a more urgent priority or dequeue.
                        # Don't bother to remove the original entry from the queue,
                        # since the vertex will be skipped when popped, and thanks to
                        # num_left_to_pick we might not need to pop it at all.
                        new_priority = old_priority - 1
                        self.priority_per_node[v] = new_priority
                        if new_priority > 0:
                            self.q.put(priority=new_priority, element=v)
                        else:
                            self.num_left_to_pick -= 1
                assert self.num_left_to_pick >= 0


class PriorityQueue:
    def __init__(self, max_priority: int) -> None:
        self.stack_per_priority: List[List[int]] = [[] for _ in range(max_priority)]

    def put(self, priority: int, element: int) -> None:
        assert priority > 0
        self.stack_per_priority[priority - 1].append(element)

    def pop(self) -> int:
        for stack in self.stack_per_priority:
            try:
                return stack.pop()
            except IndexError:
                pass
        else:
            raise ValueError("attempt to pop more than was put")
